src/inline-assembly.md

```diff
@@ -1362,6 +1384,67 @@ r[asm.rules.preserves_flags]
 > [!NOTE]
 > As a general rule, the flags covered by `preserves_flags` are those which are *not* preserved when performing a function call.
 
+r[asm.naked-rules]
+## Rules for naked inline assembly
+
+r[asm.naked-rules.intro]
+To avoid undefined behavior, these rules must be followed when using function-scope inline assembly in naked functions (`naked_asm!`):
+
+r[asm.naked-rules.reg-not-input]
+- Any registers not used for function inputs according to the calling convention and function signature will contain an undefined value on entry to the `naked_asm!` block.
+  - An "undefined value" in the context of inline assembly means that the register can (non-deterministically) have any one of the possible values allowed by the architecture.
+    Notably it is not the same as an LLVM `undef` which can have a different value every time you read it (since such a concept does not exist in assembly code).
+
+r[asm.naked-rules.reg-not-output]
+- Any callee-saved registers must have the same value upon return as they had on entry, otherwise behavior is undefined.
+  - Caller-saved registes may be used freely, even if they are not used for the return value.
```

Comment:
What's the meaning of the caveat here,

> even if they are not used for the return value.

i.e., why might someone expect that they could only be used freely only if they were used for the return value?

Reply:
well if the register is used for the return value, any manipulations that you'd perform on that register before then would be overwritten, and hence not observable from the outside.

Changes to other caller-saved registers can be observed by the caller, but the calling convention says that the caller cannot rely on the value in the registers.

So I think this just re-states a property of caller-saved registers.